Add Scene._pickVoxelCoordinate to report voxel tile and sample numbers

jjhembd opened the pull request:
<!--
Thanks for the Pull Request!

Please review [Contribution Guide](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md) before opening your first Pull Request.

To ensure your Pull Request is reviewed and accepted quickly, please refer to our [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md#pull-request-guidelines).

-->

# Description

This is another incremental step towards Voxel Picking.

A new pass type `FrameState.Passes.pickVoxel` renders an ID for the picked voxel tile and an index of the picked sample within the tile. This pass is triggered, and the rendered ID reported, by the new `Scene.pickVoxel`.

Example usage:
```javascript
  const pickedPrimitive = scene.pick(mousePosition);
  if (!Cesium.defined(pickedPrimitive)) {
    return;
  }
  if (pickedPrimitive.primitive instanceof Cesium.VoxelPrimitive) {
    const voxelCoord = scene.pickVoxel(mousePosition);
    const tileIndex = 256 * voxelCoord[0] + voxelCoord[1];
    const sampleIndex = 256 * voxelCoord[2] + voxelCoord[3];
    pickReport.innerHTML = `Tile ${tileIndex}, Sample ${sampleIndex}`;
  }
```

![image](https://github.com/CesiumGS/cesium/assets/41167620/a94c6f39-d369-40ac-ad73-51b84e81fc55)

## Testing plan

See this [testing Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=3Vltb9s4Ev4rPN99UBqvnLcuFnESXJq6WwNxE9TZbts4aGmJjolSpI+kXDuF//sOSVGiZKe57u2i6RUpJJIzw+EzL5qhE8GVRnNKPhOJjhEnn9EZUTTP4jd2Lhq1Ejs+E1xjyokctdroy4gjNMaKnOMlkYeepZ/hWyKXdjKeSJFdSjGnKZGnasmTyDAhT3tFGRng2ZDIOU1IwenpLfNvkhUsJdM4pywdiDRnxK62TpUiWnWuyELnkqjOKwxPzHpY6mm/P2ptOQH2sdWuaX1Jk09G9wlmitilWyISkdbnNM0Ig2OHc5jTDGsqeDW52uqO+IgnFs0vSCWEkzZKcEYkRisA1iFsidxrDCoTnkbFydzkG7EgrM/VjCRayAFdUG4EW3FxSsb57XAqPr+QIFddEjkksGEK0rXMiZU9yXliNIOz6AHlz0TOUxWpKZ6RLWc1OkFujI6Pj0tbm32HZvZqOSNx7/y8fzm86D8veBByB8u8RFRyngHS8Ib5fpwwwUnTYg3JG9QqLNR1z3KL+A422Yl3inlJwLC8WrbTK3Pkcg2myQRMlcLaag0MvPirwfAS/ycwGmo1wPCrFozdHftvDRJP9E2QQKQlJDWxMqT8lhETjlY9H4I1nPSUqrgACtlnt5wOnWKDeQPCALANR68IU4g5rkBLVc9IFbzRL21k/gImbmIC6K8hYQkmIE3dVIsa4LaLhaQB0TjFGlszvOmdHQS0ichmYDuur+5hOgsJ4hfnF6dX+3s3HmHnHCqRWCfTM6NJ4xBmKnLp4kEjxDMptDDax5L8JydKPwcFQGBpR+tiYmbe1YgX5vJpSIO8czInrG1f37rHO/d4b/NSwdp1bmPioWRCJ+BzpdNv8KjS29x+NbM1DNmtyMxpPIG1y/XOTbAsMU9FNiQE8prb2Sn7JJAfL+vDBdp2J9s8/TYQ7wzorGJnZJ7o13ZPg76BN6pEtK227UCprW4twMBQGVXGOkqwOYmuS/k3hnLVvSfoBjnT9AeLuYM2Mn+PK+ZKER+Y8dkzOJ0G/gMfjg9B/t0jrKn8/2PAGdT/xngzgjaF25/az8Hv3Q/raQwVphPwKs/GpugiaVRX0J91bjzLu2ANl6+BeRdISKaY80rGptCBMC6jwdJF5gxb9+Q4E8ovmMB6f+9USryMAh2f1LbzSE+AL2JEI1uAdeFxVNcW3W1vl15aUi8d9bJOvYSZgDqgXzj6RZ1+ATM1en8mRuSMyLfAtEAdFNXA/Qk+U93NDO+AYdlgWH6N4T0w3DUY7jxDxdLpOKYpkE9zAn7vFXwCRetTYPD7m/F+uf7ejvfWNp8GxjY+x6E7CJ0uWtfXBP1uvNPYal00s3X007X5pPCQsC6xfddLxaJpG0FwQFaDHdq1aqaOgxNFIUctfPZw/+4ezB8bphZrSro42/t2dDCbTbGjAWbLNWECKi8vcduAUj9MGTfX7kD1+DAcNzaLGaBkkZD/S87divNWEsK/hXev4h2znHwL675hDaAoeVfuZbXeMpQy11qGRBKsyaWkGdUUcu+s+IRCn5srDdlwimFUxK5rWWeeWAFgmZiTU8YiD3qQM0uhptRpcuI0LRqp8HaixhaV+cIrdVi+tf1SqOVhbVSSZND+M3AWSReH4aAgWLl7hEL/uubglBi+SnqIsxmDar7qyhGy1K7rn9c6fNv2D8w+8RoO9QmX3u19Ajjy8krYOg72Gc6mRJKoocy4ttr2+dQUQu7iAtpIe6hV4wO7tmvQ1ISQbepsgmVnkInEtxDg2s2ZS5pD9HEuaFquDDDl0Yti0OezXKOJss82ZBUBw+Qu++AtQSTFDBpe92KNURrez0INNJnkyiBYSIqz4vMZF6F7O+6ucdkgcRnVFVofm7c64el/n1JNHvnp5yTZj+A4W3/msIHv13IvjA/sR2KYYObuN9bvPsx6jxFzKFW/9OsxRmcKIIh//3X4y4FpT2iWZ69xSnPV/ltIbcSOuDvcELJggpVmxGSVKyHYGMsB4Xl0HcCpralGrVIufGOrPgKZOnLUKlQAtAiDWD4MGoagfHFo+lRUeMzDdy9Vin/gTsoTVkb2O1aJpJm3K+FrKRKtx3hzh5VLG+0NeJ0tmfkOyUcH19m78/6r573XjwqtZ2Lx6IB6dvH28WME/dhfBNM9tzA/GkpfyVPfHasfLVN9d8AeQ666CUsB6ChS1vxxcJiYLmY4wwnpzeE7/9IRRa54TzCfY2WEFDrT5NNrMhPS3ouIJDelgbnIKKqEZ8s+lPitimzUMszFzuYCxhZFp9YGUWUM01EY/tqlXCag/rkUilqaY+SJYsJTPx1clphNSbqhAYH5qCbL4WQu8P5RoFDczEUNGVuNOzx/c+eYG8RVpwMln9KYJ0RMNnc5jR+gipscIdOa0pZnk+blfSD4ZN917Gjv6c/QNVaSrqHF3Q7Huzc1XmW6m/u59xrc+567Mm1MoUGVL68G5yDgo70Z/NeXUqVVG9kGykwGe60+egyB4B4XtNEzuPht2PswuHjTM2dutVtHSi8ZOTHc/6aZdcFcsiiOO5qAeAgf1RnnYBEdJ0o5oP4Z+GuB+AzKRfimHqLdZFr+30+mxfHGOPl0K03P9ZPtMQ4RNBk42mmj4s9dNdgTHHW8SkcpnSOaHm/4iR0lDCsFK5Ocwdf8DgqDk6MO0NfYmMBGq4s5kQwvDcl09+TcTcZxfNSB4TqXdmUvkBuNqvkw/E7sr+MpWB3MSDnAVMopnn8A) to report tile and sample indices for random procedural voxels.

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [ ] I have updated `CHANGES.md` with a short summary of my change
- [ ] I have added or updated unit tests to ensure consistent code coverage
- [x] I have update the inline documentation, and included code examples where relevant
- [ ] I have performed a self-review of my code

ggetz commented:
Thanks for the PR @jjhembd!

Before we get into the implementation details, I wanted to check on the API. In your example, it looks like you are making two pick calls: One general `scene.pick` to determine the primitive type, and then one call to `scene.pickVoxel` to get the voxel coordinate.

Is the first call strictly necessary? Is is OK to call `scene.pickVoxel` in isolation?

jjhembd commented:
`scene.pickVoxel` currently returns tile and sample indices, but nothing to indicate _which_ voxel dataset those indices correspond to. The full sequence would look like this:
```javascript
const voxelPrimitive = scene.pick(mousePosition);
const voxelCoordinate = scene.pickVoxel(mousePosition);
const voxelData = voxelPrimitive.getData(voxelCoordinate); // This method not implemented yet
```
We could potentially combine all 3 calls into one `scene.pickVoxel` method that returns everything (or `undefined` if there is no `VoxelPrimitive` at the cursor position).

ggetz commented:
> We could potentially combine all 3 calls into one scene.pickVoxel method that returns everything (or undefined if there is no VoxelPrimitive at the cursor position).

I think that would be simplest from a public API perspective.

jjhembd commented:
I moved the coordinate picking method to a private method. We can clean up the public API in a subsequent PR.
I also tried looking up the data from CPU memory, based on the rendered tile and sample numbers. This exposed a couple bugs which are now fixed. Here is the [Sandcastle demonstrating data lookup](http://localhost:8080/Apps/Sandcastle/index.html#c=3Vptc9s4Dv4rOu99UBpXzltnduIkc6mbXj0XN5k6m20bZ1Jaoi1uKcpHUo6djv/7gaQoUZLdtHe70/QyzkgiARB8QEAA7DBlQnpzgu8x9449hu+9HhYkS4JrPeaPWqF+7qVMIsIwH7Xa3pcR87wxEvgcLTE/tCz9BE0xX+rBYMLT5JKncxJhfiqWLPQVk2dprwjFAzQbYj4nIc45Lb1m/o3TnKVgGmeERoM0yijWs61TIbAUnSu8kBnHovMWwRXRM8Rl3O+PWltGgL5stStaX5Lws9J9gqjAemqK0zCNqmOSJJjCtt0xxEiCJElZObja6o7YiIUazS+eCDHDbS9ECebIWwGwBmFNZG4DUBmzyM93Zgav0wWmfSZmOJQpH5AFYUqwFhdEeJxNh3F6/5qDXHGJ+RDDghFIlzzDWvYkY6HSDPYiB4S9TDMWCV/EaIa3jNXIxDPP3vHxcWFrte5QjV4tZzg4Oz/vXw4v+q9yHs8zG0usRK/g7AHScIfYfhDSlOG6xWqS16iVW6hrrsUSwQMsshPs5OMcg2FZOa2HV2rLxRwM4wmYKoK5VQMMtPizwbAS/ycwamrVwLCzGozdHf3XgMQSfRck4GkhjpSvDAmbUqzcUatnXbCCk4yJCHKgPH3tFsPuoVhjXofQAWzN1kvCCHyOCdBSVCNSCa//a9tTH4eJKZ8A+hsIWClNIUzdlpMS4NaTuaQBlihCEmkzXJ/1DhzaME1mYDsmrzYw9VyC4PX5xenV/t6tRdgcDhFyJMO4pzSpbUIN+SZcPGqEYMZTmSrtA47/nWEhX4ECILCwoz5i6UzdixHLzWXDkAR553iOaVvfvjeXD+byUcelnLVrjo3yh4LJO4EzVxz6NSeqOG1mvYrZaobslmRqN5ZA2+Vm59aZ5ohFaTLEGOKaWdko+8yRHyyrjwtv2+xs/fB7R7wxoLGKHuFZKN/pNRX6Cl6/FNHW2rYdpba6FQcDQyVEKOuIlM6xf1PIv1WUq+4GpxtkVJKfzOcO2p76PC2fK0TcUXVme7A7CfwH1h0fg/yHe1hd+f9Hh1Oo/4X+pgStc7f/aj0Dvz1+SMYBZJhGwNssGaukC0d+VUG717k6WfYIVnD5GpgPjoQwRoyVMta5Drhx4Q2azld72NoQ45Qrv6Ypkvt7p5yjpe/o+KyynEV6Anw+xdLTCVgXLkdVbb2H7e3ilBbUS0O9rFIvYcShdugXhn5RpV/ASIXe7oliPsP8PTAtvI7nV8B9Dq+p7nqGD8CwrDEsv8bwERgeagwPlqFk6XQMUwzkcYbh3FsFn0HS+gIY7Prqeb+Y/6if9xqLx46x1ZljUB24h85v6qucfjfYqS3VFE11Hv2iMR7mJ8TNS3Td9UZQP2574BwQ1WCFdiWbqeJgRBGIUQsbPczfw6PxY83QoqGk8bO970cH0VmMDA0wa64JTSHzshK3FSjVzRR+c2M2VPUPxXGro5gCiucB+Rs5d0vOKceYfQ/vXsk7phn+HtZ9xepAUfCuzM2qWTIUMhslQ8gxkviSk4RIArF3lr9Coc7NhIRoGCN4yn3XlKwzSywAsCSd41NKfQu6EzMLoSrVqXOiKMoLKbc7UWHzi3hhlTos7tp2ytXysPJUkCRQ/lM4LJwsDt2HnGBl+gi5/lXN4VAieCvJIUpmFLL5sipvUP5BpNTdFt090BSayvQF5pUegG4MDJQmQQOp6oB5AeiOAxz15VWqMz3QZDiLMcd+TYlxZbZtI65KlUxrAwpNve1V7RXcWNUpe1xQ19U+zrQx2YSjKYQAacZUG+fQ+zRPSVTMDBBh/uv8oc9mmfQmQl/bEHdSeAwfkjtrK8wJolASmxttruJo2FHIkiaTTCgEc0lBkr9gg9y5p+Nug0u70WYeZPKzT/VmkAvJ7zGR+IlDMsfhvg9Rf2szArt5C6K5WcdlKiEbng/0u2UYImraIs2WiZo/o1htSlR7hWeUkpkACILf/zn89UBVNSTJkncoIplo/yWk2tFHzGxuCMEzREJSrILRVZrSMeIDzDL/xoFTalONWoVceDWX5Yen0s9RK1cB0MIUHPzQqTOcrMegaSNYfmIeb9mUb4ZHWlmWsDSyXbGMLvVwXwpvRFav6fj1FVYmlrTX4NVbUvX64k8Ort6H8/7bV2fvnhRaL9PFkwPq5cX7p48RlHF/Ekwbmjc/G0pfiVM/HKufLVL9cMCeQqy6dVMBKEQiWv9OcRiq4mc4QyE+m8N7/o0h8k3OHyI2R0IJyXVW39BFPajaIFFFOnOK0jBTCYLqguS5wstlH+qDVp141GoKMgnp4zJ0K1Ox55tQLSCdX51qc/qlXVVNowRU2oJJCqnUZSqIpjn2LFGAWWSHnXaNWfVi/AecmbL+gUG/IsjgrfqHf8vRzBuDvitgq9Y/tF3DjeWWy1xWXeVatboBEkwhEQtxOllfin2TAtpMxV7vlA7Xxbg237q95+1M8I2+aTh4ey9eQNFbiryBCn3bfd51O51ClWabefdqvPuGt36wAgIVNn9zNTgHIZ90a/PvXwqlVm1PV4Bq0Flv9cnR4zNeTtRXuG8hW26UccGd5GiOuYCEG8wb/csh9otlXEBsIQKiXMlFgVLv907H6BrRTPfZiyJGUBJi30Xomer21wYAoAN36XDT91tBEJTrbFVwVMWSkEvIpcco/Dzlqg4tv5FRszLtCaGHhlA5sKnuGACwG4KIjn+Di9+GZ3eDi+szRdxqt470Gidq5X+QZJZy6WWcgmIdiWFPYEjRGWegkgxCkav4i9LxHdbE+SEudXyulTvU+Pk7bS//mCZSfsR/aYQsW1YTAWsuD8F91K8Jno9pGn7OA+gMagrY5aG3G8bF/34YrxGrUPoGifckkjHIw7bhE2MyjaU70tzXvSpOiy+e+ZSw5xRPpFbLqnLUsageRWTukeh4zc9DvJAiIWBmklFIKR8gFJ8cdYC+wkZTpHZ9AUcdtqFI4t2TczMIp+eoA49NLmlqLyBXGpXjpeHyqfpk9dVwon/8EcF5Kwxl11rPqt8IJyVNoVl+/Q8).

Remaining tasks for this PR:

- [ ] Add spec for new `Scene._pickVoxelCoordinate` method
- [ ] Address TODO items in comments

ggetz commented on packages/engine/Source/Scene/Picking.js:
```diff
@@ -236,6 +236,19 @@ let scratchRectangle = new BoundingRectangle(
 const scratchColorZero = new Color(0.0, 0.0, 0.0, 0.0);
 
+/**
+ * Returns an object with a `primitive` property that contains the first (top) primitive in the scene
```
```suggestion
 * Returns an object with a <code>primitive</code> property that contains the first (top) primitive in the scene
```

ggetz commented on packages/engine/Source/Scene/Picking.js:
```diff
@@ -301,6 +314,82 @@ Picking.prototype.pick = function (scene, windowPosition, width, height) {
+Picking.prototype.pickVoxel = function (scene, windowPosition, width, height) {
+  //>>includeStart('debug', pragmas.debug);
+  if (!defined(windowPosition)) {
+    throw new DeveloperError("windowPosition is undefined.");
```
Why not `Check.defined`?

jjhembd replied:
Good point. This was a copy-paste from `Picking.prototype.pick`. I switched to `Check.defined` in both places.

ggetz commented on packages/engine/Source/Scene/VoxelTraversal.js:
```diff
@@ -224,7 +224,7 @@ function VoxelTraversal(
  */
 VoxelTraversal.prototype.findKeyframeNode = function (megatextureIndex) {
   return this._keyframeNodesInMegatexture.find(function (keyframeNode) {
-    return (keyframeNode.megatextureIndex = megatextureIndex);
```
This seems like it could have caused some problems... Any idea if it fixed existing bugs?

jjhembd replied:
I introduced this in https://github.com/CesiumGS/cesium/pull/11754, and didn't notice the bug until I did more testing in this PR.

ggetz commented:
@jjhembd Updates are looking good to me! Anything else in this PR before merging?

jjhembd commented:
@ggetz I think this one is ready to go. The remaining changes will go in the 3rd PR.
Thanks for your reviews!

lilleyse commented:
@jjhembd it looks like `frameState.passes.pick` and `frameState.passes.pickVoxel` are mutually exclusive but it might simplify the code if they are both true.

```js
  frameState.passes.pick = true;
  frameState.passes.pickVoxel = true;
```

That way the code can leverage the existing picking infrastructure and only check for `frameState.passes.pickVoxel` in `VoxelPrimitive`.

@javagl and I were discussing this in the context of https://github.com/CesiumGS/cesium/pull/12075 for a new flag `frameState.passes.pickMetadata`.
